comparison (#2329)
JuliaData/DataFrames.jl: docs/src/man/comparisons.md

> The table compares the main functions of DataFrames with R dplyr and Stata

Reviewer: which versions of R/dplyr and Stata? If we do it also Pandas should be added I think.

Reviewer: also later I would add DataFramesMeta.jl (when @pdeffebach is done with the work he is concentrating on now)

Reviewer: also add that `df` is a data frame here.

Author: I do not know Panda unfortunately. Maybe someone could add it after?

> The table compares the main functions of DataFrames with R dplyr and Stata

Reviewer: ```suggestion
The table compares the main functions of DataFrames.jl with R dplyr and Stata
```

| DataFrames       | dplyr | Stata|
|:------------|:------------|:------------|

> |`combine(df, :x => mean)`  | `summarize(mean(x))`    | `collapse (mean) x `|

Reviewer: I think that in dplyr functions taka a data frame as a first argument

> |`combine(df, :x => mean)`  | `summarize(mean(x))`    | `collapse (mean) x `|

Reviewer 2: I don't think ungrouped sumaries are very common. Better to show off `describe` from data frames and `summary` from stata?

Author: Adding `groupby` for each command makes it too verbose. To clarify, I've re-arranged the table after the paragraph explaining how to use `groupby`.

|`transform(df, :x => mean)`   | `mutate(mean(x))`    | `egen x_mean = mean(x)`|
|`select(df, :x, :y)`   | `select(df, x, y)`  | `keep x y` |
|`sort(df, :x)`   | `arrange(df, x)`    | `sort x`|

> |`filter(:x => x -> x >= 1, df)`   | `filter(df, x >= 1)`  | `keep if x >= 1` |

Reviewer: ```suggestion
|`filter(:x => >=(1), df)`   | `filter(df, x >= 1)`  | `keep if x >= 1` |
```

Author: I understand that it is a shorthand but it seems a bit harder to understand, and less generalizable, so I'm not sure I want to change it

Reviewer: OK

Here are more complicated examples for `combine`/`transform`/`select`:
| DataFrames       | dplyr | Stata|
|:------------|:------------|:------------|
|`combine(df, :x => mean => :x_mean)`   | `summarize(df, x_mean = mean(x))`    | `collapse (mean) x_mean = x`|

> |`combine(df, :x => x -> maximum(x) - minimum(x))`   | `summarize(df, max(x) - min(x))`    | |

Reviewer: you want to say that this is not achievable in Stata at all - if this is the case it is strange (but I do not know Stata)

Reviewer: also - out of couriosity - what would be the equivalent of:
```
combine(:x => x -> (extremum=["max", "min"], value=[maximum(x), minimum(x)]), df)
```
in dplyr and Stata?

Author: In Stata, `combine` only accepts a specific set of functions. To  do more complicated things, you would have to do a combination of `transform!`and `filter!`:
```
egen out = cov(x, y)
keep if _n == 1
```

Author: added

|`combine(df, [:x, :y] => cov)`   | `summarize(df, cov(x, y))`    | |
|`combine(df, :x => maximum,  :y => minimum)`   | `summarize(df, max(x), min(y))`    | `collapse (max) x (min) y` |
|`combine(df, [:x, :y] .=> mean)`   | `summarize(df, across(c(x, y), mean))`    | `collapse (mean) x y` |

> |`combine(df, names(df, r"^x.*") .=> mean)`   | `summarize(df, across(starts_with("x"), mean))`    | `collapse (mean) x*` |

Reviewer: ```suggestion
|`combine(df, names(df, r"^x") .=> mean)`   | `summarize(df, across(starts_with("x"), mean))`    | `collapse (mean) x*` |
```

> |`combine(df, names(df, r"^x.*") .=> mean)`   | `summarize(df, across(starts_with("x"), mean))`    | `collapse (mean) x*` |

Reviewer 2: again, not sure how frequent `combine` without grouping is

> As in dplyr and, the functions `select`, `transform`, `combine` can also be used on grouped dataframes:

Reviewer: Stata is missing somewhere here I think.

| DataFrames       | dplyr | Stata|
|:------------|:------------|:------------|
|`transform(groupby(df, :id), :x => mean)`   | `mutate(group_by(df, id), mean(x))`    | `egen x_mean = mean(x), by(id)`|

> |`combine(groupby(df, :id), :x => mean)`  | `summarize(group_by(df, id), mean(x))`    | `collapse (mean) x, by(id)`|

Reviewer: I would put the functions so that `combine` is first or last, as it does a different thing than `select` and `transform`.

|`select(groupby(df, :id), :x => mean)`   | `transmute(group_by(df, id), mean(x))`    | |

> 

Reviewer: I would add at the end that we support `select!` and `transform!` that work in place. How does that compare to dplyr (I guess it is not possible) and  Stata (here maybe this is what happens by default?).

| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|
|`combine(groupby(df, :id), :x => mean)`|`summarize(group_by(df, id), mean(x))`|`collapse (mean) x, by(id)`|
|`transform(groupby(df, :id), :x => mean)`|`mutate(group_by(df, id), mean(x))`|`egen x_mean = mean(x), by(id)`|
|`select(groupby(df, :id), :x => mean)`|`transmute(group_by(df, id), mean(x))`||

The table compares more complicated syntaxes 
Operations| DataFrames       | dplyr| Stata|
|:------------|:------------|:------------|:------------|
|Several columns |`combine(df, :x => maximum,  :y => minimum)`|`summarize(df, max(x), min(y))`|`collapse (max) x (min) y`|
||`combine(df, [:x, :y] .=> mean)`|`summarize(df, across(c(x, y), mean))`|`collapse (mean) x y`|
||`combine(df, names(df, r"^x") .=> mean)`|`summarize(df, across(starts_with("x"), mean))`|`collapse (mean) x*`|
|Multivariate function|`combine(df, [:x, :y] => cov)`|`summarize(df, cov(x, y))`||
|DataFrame as output|`combine(:x => x -> (name = ["minimum", "maximum"], value = [minimum(x), maximum(x)]), df)`|`summarize(df, tibble(name = c("minimum", "maximum"), value = range(x)))`||

> |DataFrame as input|`combine(d -> head(d, 1), df)`|`summarize(df, head(across(), 1))`||

Reviewer: we do not have `head`. use `first`.

|`transform(groupby(df, :id), :x => mean)`   | `mutate(group_by(df, id), mean(x))`    | `egen x_mean = mean(x), by(id)`|
|`select(groupby(df, :id), :x => mean)`   | `transmute(group_by(df, id), mean(x))`    | |

Here are more complicated examples to work with multiple functions / multiple columns
| DataFrames       | dplyr | Stata|
|:------------|:------------|:------------|
|`combine(df, :x => mean => :x_mean)`   | `summarize(df, x_mean = mean(x))`    | `collapse (mean) x_mean = x`|
|`combine(df, :x => x -> maximum(x) - minimum(x))`   | `summarize(df, max(x) - min(x))`    | |
|`combine(df, [:x, :y] => cov)`   | `summarize(df, cov(x, y))`    | |
|`combine(df, :x => maximum,  :y => minimum)`   | `summarize(df, max(x), min(y))`    | `collapse (max) x (min) y` |
|`combine(df, [:x, :y] .=> mean)`   | `summarize(df, across(c(x, y), mean))`    | `collapse (mean) x y` |
|`combine(df, names(df, r"^x") .=> mean)`   | `summarize(df, across(starts_with("x"), mean))`    | `collapse (mean) x*` |
|`combine(:x => x -> (name = ["minimum", "maximum"], value = [minimum(x), maximum(x)]), df)`   | `summarize(df, tibble(name = c("minimum", "maximum"), value = range(x)))`    | |

> |`combine(d -> head(d, 1), df)` | `summarize(df, head(across(), 1))`||

Reviewer: we do not have `head` function in DataFrames.jl. Just use `first` as @pdeffebach suggested. If you want more than one row use `first(d, 2)`. `first` is the function in Base for this operation.

if you want to show a general row selection use e.g. `d -> d[1:2, :]`. The only difference is that this will fail when data frame has less than two rows, and `first(d, 2)` will work.

The table compares the main functions of DataFrames.jl with the R package dplyr (v1) and Stata (>=v8)
|Operations| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|:------------|
|Reduce multiple values|`combine(df, :x => mean)`|`summarize(df, mean(x))`|`collapse (mean) x =`|
|Add new columns|`transform(df, :x => mean => :x_mean)`|`mutate(df, x_mean = mean(x))`|`egen x_mean = mean(x)`|
|Pick columns|`select(df, :x, :y)`|`select(df, x, y)`|`keep x y`|

> |Pick rows |`filter(:x => x -> x >= 1, df)`|`filter(df, x >= 1)`|`keep if x >= 1`|

Reviewer: I was thinking about it - if you want to show how to filter using a custom defined predicate in Julia then probably it should be shown in R and Stata how to define such a predicate. Otherwise in Julia you should write `>=(1)` as I have suggested earlier I think. Or maybe show both in two consecutive lines?

Author: I don't think it's super easy to understand your suggested syntax. You have to know that >=(1) creates a Function which is really not obvious. I changed the predicate to ismissing to sidestep this.

Reviewer: But this is exactly the point - we want to teach readers here that there is such an option I think; otherwise we will always have a problem that they do not know this syntax, because we do not show it :smile: .
I would keep `ismissing` example as it is nice

Reviewer 3: Maybe there can be another line below showing both `:x => x -> x > 1` and `:x => >(1)` in the same cell? I agree with @bkamins that it's a good occasion to show people how to do this. If we only show `ismissing`, people won't even now how to do more complex operations.

|Sort rows|`sort(df, :x)`|`arrange(df, x)`|`sort x`|
|Rename columns|`rename(df, :x => :v)`|`rename(df, v = x)`|`rename x v`|

These functions create new dataframes (like in dplyr). To mutate dataframes in place (like in Stata), use the suffix `!` (e.g. `transform!`, `select!`, etc)

The functions `select`, `transform`, `combine` can be applied on grouped dataframes:
| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|
|`combine(groupby(df, :id), :x => mean)`|`summarize(group_by(df, id), mean(x))`|`collapse (mean) x, by(id)`|
|`transform(groupby(df, :id), :x => mean)`|`mutate(group_by(df, id), mean(x))`|`egen x_mean = mean(x), by(id)`|
|`select(groupby(df, :id), :x => mean)`|`transmute(group_by(df, id), mean(x))`||

The table compares more complicated syntaxes 
Operations| DataFrames       | dplyr| Stata|
|:------------|:------------|:------------|:------------|
|Transform several columns |`combine(df, :x => maximum,  :y => minimum)`|`summarize(df, max(x), min(y))`|`collapse (max) x (min) y`|

> ||`combine(df, [:x, :y] .=> mean)`|`summarize(df, across(c(x, y), mean))`|`collapse (mean) x y`|

Reviewer: also recently several people asked about `([:x, :y] .=> [mean std length])...`, maybe also add this?

Author: Added

Reviewer: you have added something else :smile:, but I think the example is also nice

||`combine(df, names(df, r"^x") .=> mean)`|`summarize(df, across(starts_with("x"), mean))`|`collapse (mean) x*`|
|Multivariate function|`transform(df, [:x, :y] => cov)`|`mutate(df, cov(x, y))`|`egen z = corr(x y)`|
|Row-wise|`transform(df, [:x, :y] => ByRow(min))`|`mutate(rowwise(df), min(x, y))`|`egen z = rowmin(x y)`|

> ||`transform(df, AsTable(r"^x") => ByRow(sum))`|`mutate(rowwise(df), sum(c_across(starts_with("x"))))`|`egen z = rowtotal(x*)`|

Reviewer: I do not remember if we discussed this but you can also write `r"^x" => ByRow(+)`, maybe show both options?

Author: I think the one I have now is the more straightforward comparison with dplyr.

Reviewer: I agree this is a straightforward comparison, but again - as written above, normally I would write `r"^x" => ByRow(+)` in my code, or even `r"^x" => +` in an interactive session, as this is much convenient to write. So the question is if we want to show users here typical patterns that work (I agree there is a slight difference between `+` and `sum` so if you feel `+` is too confusing I will not insist :smile: ).

Author: Ok so to sidestep this I use the function arg max as an example.

> ||`transform(df, AsTable(names(df, r"^x")) => ByRow(sum))`|`mutate(rowwise(df), sum(c_across(starts_with("x"))))`|`egen z = rowtotal(x*)`|

Reviewer: ```suggestion
||`transform(df, AsTable(r"^x") => ByRow(sum))`|`mutate(rowwise(df), sum(c_across(starts_with("x"))))`|`egen z = rowtotal(x*)`|
```
or shorter:
```
transform(df, r"^x" => ByRow(+))
```

> The table compares the main functions of DataFrames.jl with the R package dplyr (v1) and Stata (>=v8)

Reviewer 3: ```suggestion
The following table compares the main functions of DataFrames.jl with the R package dplyr (version 1) and Stata (version 8 and above)
```

Also you'll have to add a heading at the top of the file.

|Operations| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|:------------|
|Reduce multiple values|`combine(df, :x => mean)`|`summarize(df, mean(x))`|`collapse (mean) x =`|
|Add new columns|`transform(df, :x => mean => :x_mean)`|`mutate(df, x_mean = mean(x))`|`egen x_mean = mean(x)`|

> |Pick columns|`select(df, :x, :y)`|`select(df, x, y)`|`keep x y`|

Reviewer 3: I'd also show how to compute new columns using `select`. This is `transmute` in dplyr.

|Pick rows |`filter(:x => ismissing, df)`|`filter(df, is.na(x))`|`keep if missing(x)`|
|Sort rows|`sort(df, :x)`|`arrange(df, x)`|`sort x`|

> |Rename columns|`rename(df, :x => :v)`|`rename(df, v = x)`|`rename x v`|

Reviewer 3: How about putting this after "adding new columns"? That sounds more logical.

> These functions create new dataframes (like in dplyr). To mutate dataframes in place (like in Stata), use the suffix `!` (e.g. `transform!`, `select!`, etc)

Reviewer 3: ```suggestion
These functions create new data frames (like in dplyr). To mutate data frames in place (like in Stata), use the suffix `!` (e.g. `transform!`, `select!`, etc)
```

> The functions `select`, `transform`, `combine` can be applied on grouped dataframes:

Reviewer 3: ```suggestion
The functions `select`, `transform` and `combine` can be applied on grouped data frames, in which case they operate by group:
```

> | DataFrames.jl       | dplyr | Stata|

Reviewer 3: Could you add a column describing the operation like in the previous table?

|:------------|:------------|:------------|
|`combine(groupby(df, :id), :x => mean)`|`summarize(group_by(df, id), mean(x))`|`collapse (mean) x, by(id)`|
|`transform(groupby(df, :id), :x => mean)`|`mutate(group_by(df, id), mean(x))`|`egen x_mean = mean(x), by(id)`|
|`select(groupby(df, :id), :x => mean)`|`transmute(group_by(df, id), mean(x))`||

> The table compares more complicated syntaxes 

Reviewer 3: ```suggestion
Finally, the table below compares more complicated syntaxes:
```

Operations| DataFrames       | dplyr| Stata|
|:------------|:------------|:------------|:------------|
|Transform several columns |`combine(df, :x => maximum,  :y => minimum)`|`summarize(df, max(x), min(y))`|`collapse (max) x (min) y`|
||`combine(df, [:x, :y] .=> mean)`|`summarize(df, across(c(x, y), mean))`|`collapse (mean) x y`|

> ||`combine(df, [:x, :y] => [maximum minimum])`|`summarize(df, across(c(x, y), list(max, min)))`|`collapse (max) x y (min) x y`|

Reviewer: ```suggestion
||`combine(df, [[:x, :y]] .=> [maximum minimum])`|`summarize(df, across(c(x, y), list(max, min)))`|`collapse (max) x y (min) x y`|
```

||`combine(df, names(df, r"^x") .=> mean)`|`summarize(df, across(starts_with("x"), mean))`|`collapse (mean) x*`|

> |Multivariate function|`transform(df, [:x, :y] => cov)`|`mutate(df, cov(x, y))`|`egen z = corr(x y)`|

Reviewer 3: ```suggestion
|Multivariate function|`transform(df, [:x, :y] => cor)`|`mutate(df, cor(x, y))`|`egen z = corr(x y)`|
```

# Comparisons with other languages

> The following table compares the main functions of DataFrames.jl with the R package dplyr (version 1) and Stata (version 8 and above)

Reviewer 2: ```suggestion
The following table compares the main functions of DataFrames.jl with the R package dplyr (version 1) and Stata (version 8 and above)

```

You need a new line for the html write up to work

|Rename columns|`rename(df, :x => :v)`|`rename(df, v = x)`|`rename x v`|
|Pick columns|`select(df, :x, :y)`|`select(df, x, y)`|`keep x y`|
|Pick & transform columns|`select(df, :x => mean, :y)`|`transmute(df, mean(x), y)`||
|Pick rows |`filter(:x => >=(1), df)`|`filter(df, x >= 1)`|`keep if x >= 1`|
|Sort rows|`sort(df, :x)`|`arrange(df, x)`|`sort x`|

These functions create new data frames (like in dplyr). To mutate data frames in place (like in Stata), use the suffix `!` (e.g. `transform!`, `select!`, etc)

The functions `select`, `transform` and `combine` can be applied on grouped data frames, in which case they operate by group:
|Operations| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|:------------|
|Reduce multiple values|`combine(groupby(df, :id), :x => mean)`|`summarize(group_by(df, id), mean(x))`|`collapse (mean) x, by(id)`|
|Add new columns|`transform(groupby(df, :id), :x => mean)`|`mutate(group_by(df, id), mean(x))`|`egen x_mean = mean(x), by(id)`|
|Pick & transform columns|`select(groupby(df, :id), :x => mean)`|`transmute(group_by(df, id), mean(x))`||

> Finally, the table below compares more complicated syntaxes:

Reviewer 2: ```suggestion
Finally, the table below compares more complicated syntaxes:

```

|Sort rows|`sort!(df, :x)`|`sort x`|

Note that the suffix `!` (i.e. `transform!`, `select!`, etc) ensures that the operation transforms the dataframe in place, as in Stata

Some of these functions can be applied to grouped data frames, in which case they operate by group:
|Operations| DataFrames.jl     | Stata|
|:------------|:------------|:------------|
|Add new columns|`transform!(groupby(df, :id), :x => mean)`|`egen x_mean = mean(x), by(id)`|
|Reduce multiple values|`combine(groupby(df, :id), :x => mean)`|`collapse (mean) x, by(id)`|

The table below compares more complicated syntaxes:

Operations| DataFrames.jl | Stata|
|:------------|:------------|:------------|
|Transform certain rows |`transform(df, [:x, :y] => ((x, y)-> ifelse.(y .>= 1, 1, x)) => :x)`|`replace x = 1 if y >= 1`|
|Transform several columns |`combine(df, :x => maximum,  :y => minimum)`|`collapse (max) x (min) y`|
||`combine(df, [:x, :y] .=> mean)`|`collapse (mean) x y`|
||`combine(df, names(df, r"^x") .=> mean)`|`collapse (mean) x*`|
||`combine(df, ([:x, :y] .=> [maximum minimum])...)`|`collapse (max) x y (min) x y`|
|Multivariate function|`transform!(df, [:x, :y] => cor => :z)`|`egen z = corr(x y)`|

> |Row-wise|`transform!(df, [:x, :y] => ByRow(min) => :z)`|`egen z = rowmin(x y)`|

Reviewer 3: Maybe it would be useful to add an example of a simple row-wise function using `gen` in Stata? For example it's not obvious that `ByRow` is also useful for things like `:x => ByRow(x -> x^2) => x²`.

# Comparisons with other languages

The following table compares the main functions of DataFrames.jl with the R package dplyr (version 1) and Stata (version 8 and above)
|Operations| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|:------------|
|Reduce multiple values|`combine(df, :x => mean)`|`summarize(df, mean(x))`|`collapse (mean) x =`|
|Add new columns|`transform(df, :x => mean => :x_mean)`|`mutate(df, x_mean = mean(x))`|`egen x_mean = mean(x)`|
|Rename columns|`rename(df, :x => :v)`|`rename(df, v = x)`|`rename x v`|
|Pick columns|`select(df, :x, :y)`|`select(df, x, y)`|`keep x y`|

> |Pick & transform columns|`select(df, :x => mean)`|`transmute(df, mean(x), y)`||

Reviewer 3: ```suggestion
|Pick & transform columns|`select(df, :x => mean, :y)`|`transmute(df, mean(x), y)`||
```

|Pick rows |`filter(:x => >=(1), df)`|`filter(df, x >= 1)`|`keep if x >= 1`|
|Sort rows|`sort(df, :x)`|`arrange(df, x)`|`sort x`|

These functions create new data frames (like in dplyr). To mutate data frames in place (like in Stata), use the suffix `!` (e.g. `transform!`, `select!`, etc)

The functions `select`, `transform` and `combine` can be applied on grouped data frames, in which case they operate by group:
|Operations| DataFrames.jl       | dplyr | Stata|
|:------------|:------------|:------------|:------------|
|Reduce multiple values|`combine(groupby(df, :id), :x => mean)`|`summarize(group_by(df, id), mean(x))`|`collapse (mean) x, by(id)`|
|Add new columns|`transform(groupby(df, :id), :x => mean)`|`mutate(group_by(df, id), mean(x))`|`egen x_mean = mean(x), by(id)`|

> |Pick columns|`select(groupby(df, :id), :x => mean)`|`transmute(group_by(df, id), mean(x))`||

Reviewer 3: ```suggestion
|Pick & transform columns|`select(groupby(df, :id), :x => mean)`|`transmute(group_by(df, id), mean(x))`||
```
